### PR TITLE
Fixed scenarios where AI didn't challenge assassin or bluffed contessa when left with last influence

### PR DESCRIPTION
**Pull request by einsteinsfool**
All 69 tests passed.

### CODE REVIEW

File: ai-player.js [hunk 1]
```diff
@@ -235,10 +235,46 @@ function createAiPlayer(game, options) {
         if (state.state.name == stateNames.FINAL_ACTION_RESPONSE) {
             return false;
         }
+
+        // Challenge if somebody claims to have role that was revealed 3 times or we have the rest of them
+        var usedRoles = countRevealedRoles(state.state.action);
```
Comment (octachrome) [inline ai-player.js]: This seems wrong: `state.state.action` is an action, e.g., `assassinate`, but `countRevealedRoles` accepts a role name, e.g., `assassin`. Also, on the line below, you are comparing a role with an action.

File: ai-player.js [hunk 2]
```diff
@@ -235,10 +235,46 @@ function createAiPlayer(game, options) {
         if (state.state.name == stateNames.FINAL_ACTION_RESPONSE) {
             return false;
         }
+
```
Comment (octachrome) [inline ai-player.js]: This function is called in two different places and I think you need to handle them differently:

- An opponent has played an action and you are deciding whether to challenge him (your code handles this case)
- You have played an action, your opponent has blocked the action, and you are deciding whether to challenge him (I think your code would make an incorrect choice in this case. It should be enough to wrap your checks with an `if (state.state.name == stateNames.ACTION_RESPONSE)`)

File: ai-player.js [hunk 3]
```diff
@@ -235,10 +235,46 @@ function createAiPlayer(game, options) {
...
+            return true;
+        }
+
+        // Challenge if you're being assassinated, it's your last influence and all contessas have been used
+        if (state.state.action === 'assassinate' && state.players[state.playerIdx].influenceCount === 1) {
+            var contessas = countRevealedRoles('contessa');
+            if (contessas === 3) {
+                return true;
+            }
+            // Challenge when there are no contessas that were not claimed
```
Comment (octachrome) [inline ai-player.js]: This line took me a while to understand. Could you update the comment to explain that you are counting all contessas that have been revealed or claimed?

**Comment by einsteinsfool** [comment]
I think I fixed all issues that you mentioned and I found additional bugs in `countRevealedRoles`.
Thanks for reporting them.
The PR status is still "changes requested", though. I don't think any more changes are needed but it you disagree, feel free to comment.

Also, I've noted 9 more things to change in AI script and it would be easier if I could choose starting roles for all AIs before match. Can I do that somehow?

**Comment by octachrome** [comment]
Not sure what you mean by "choose starting roles for all AIs before match" - players do not get dealt their roles until the game begins. Can you explain more what you mean?

**Comment by einsteinsfool** [comment]
Well, in that case I want to be able to modify AIs' influences after game starts. Is it possible?
So, how do you test a speific scenario? E.g. you want to check what AI does when it revealed contessa, has another contessa and the opponent also has contessa and an assassin.
For now I restart the match until I get specific roles that is a bit boring.

**Comment by octachrome** [comment]
Look at the AI unit tests. They assign specific roles and states and then they verify how the AI reacts. If you are changing AI behaviour I strongly recommend adding unit tests for it.

**Comment by einsteinsfool** [comment]
I'll give that a go, thanks. :-)

**Pull request merged**